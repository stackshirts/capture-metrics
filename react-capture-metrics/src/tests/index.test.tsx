import React, { useRef, useState } from 'react'
import * as rtl from '@testing-library/react';
import { MetricsProvider, useMetrics } from '../index';

const analytics = {
  page: jest.fn(),
  track: jest.fn(),
}

const CallTrackOnUpdate: React.FC = () => {
  const { analytics } = useMetrics()
  const ref = useRef(0);

  return (
    <button
      onClick={() => {
        ref.current += 1;
        analytics.track('updated', { count: ref.current })
      }}
    >
      Child
    </button>
  );
}

const Test: React.FC = ({ children }) => {

  const [parentCount, setParentCount] = useState(0);

  return (
    <MetricsProvider analytics={analytics} properties={{ parentCount }}>
      <button onClick={() => setParentCount(parentCount + 1)}>
        Provider
      </button>
      <div data-testid="parentCount">
        {parentCount}
      </div>

      {children}
    </MetricsProvider>
  )
}

describe('props updated when changed in MetricsProvider', () => {
  it('works', () => {

    rtl.render(
      <Test>
        <CallTrackOnUpdate />
      </Test>
    );

    rtl.fireEvent.click(rtl.screen.getByText(/Child/i))

    expect(analytics.track.mock.calls[0]).toMatchSnapshot();

    rtl.fireEvent.click(rtl.screen.getByText(/Provider/i))
    rtl.fireEvent.click(rtl.screen.getByText(/Child/i))
    expect(rtl.screen.getByTestId(/parentCount/i).textContent).toEqual('1')

    expect(analytics.track.mock.calls[1]).toMatchSnapshot();

  })
})
