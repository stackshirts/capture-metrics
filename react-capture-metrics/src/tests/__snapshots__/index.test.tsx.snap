// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageView will capture props in useMetrics and on itself works 1`] = `
Array [
  "updated",
  Object {
    "TestPageView": "props",
    "TestReady": "props",
    "count": 1,
    "pageCategory": "Category",
    "pageName": "Page",
    "parentCount": 0,
  },
]
`;

exports[`props updated when changed in MetricsProvider works 1`] = `
Array [
  "updated",
  Object {
    "count": 1,
    "parentCount": 0,
  },
]
`;

exports[`props updated when changed in MetricsProvider works 2`] = `
Array [
  "updated",
  Object {
    "count": 2,
    "parentCount": 1,
  },
]
`;
